segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1350 @@
+                txn.commit(!didSegmentLayoutChange);
+                isCommited = true;
+
+                // Post commit actions.
+                if (isStorageSystemSegment(streamSegmentName) && chunksAddedCount > 0) {
+                    // commit all system log records.
+                    Preconditions.checkState(chunksAddedCount == systemLogRecords.size());
+                    systemJournal.commitRecords(systemLogRecords);
+                }
+
+                // Update the read index.
+                val readIndex = getReadIndex(streamSegmentName);
+                for (val entry : newReadIndexEntries.entrySet()) {
+                    readIndex.put(entry.getKey(), entry.getValue());
+                }
+
+                Duration elapsed = timer.getElapsed();
+                log.debug("write - segment={} offset={} length={} latency={}.", handle.getSegmentName(), offset, length, elapsed.toMillis());
+                LoggerHelpers.traceLeave(log, "write", traceId, handle, offset);
+                return null;
+            } catch (StorageMetadataVersionMismatchException ex) {
+                throw ex;
+            } catch (StorageMetadataWritesFencedOutException ex) {
+                throw new StorageNotPrimaryException(streamSegmentName, ex);
+            } catch (Exception ex) {
+                throw ex;
+            } finally {
+                if (!isCommited && chunksAddedCount > 0) {
+                    // Collect garbage.
+                    collectGarbage(newReadIndexEntries.values());
+                }
+            }
+        });
+    }
+
+    /**
+     * Gets whether given segment is a critical storage system segment.
+     * @param streamSegmentName Name of the segment.
+     * @return True if this is a storage system segment.
+     */
+    private boolean isStorageSystemSegment(String streamSegmentName) {
+        return null != systemJournal && systemJournal.isStorageSystemSegment(streamSegmentName);
+    }
+
+    /**
+     * Adds a system log.
+     * @param systemLogRecords
+     * @param streamSegmentName Name of the segment.
+     * @param offset Offset at which new chunk was added.
+     * @param oldChunkName Name of the previous last chunk.
+     * @param newChunkName Name of the new last chunk.
+     */
+    private void addSystemLogRecord(ArrayList<String> systemLogRecords, String streamSegmentName, long offset, String oldChunkName, String newChunkName) {
+        val systemLogRecord = systemJournal.getChunkAddedRecord(streamSegmentName,
+            offset,
+            oldChunkName == null ? null : oldChunkName,
+            newChunkName);
+        systemLogRecords.add(systemLogRecord);
+    }
+
+    /**
+     * Delete the garbage chunks.
+     * @param chunksTodelete List of chunks to delete.
+     */
+    private void collectGarbage(Collection<String> chunksTodelete) {
+        for (val chunkTodelete : chunksTodelete) {
+            try {
+                if (chunkStorage.exists(chunkTodelete)) {
```

Comment:
what about a deleteIfExists primitive on checkStorage ?
sometimes implementations can be more efficient.

Therefore there is no lock/transaction here, so there can be a race, and two "threads" can issue the "delete" operation.

Maybe you can simply call chuckStorage.delete and require it to not fail in case of non existant chunk

Reply:
If object doesn't exist then code has to throw ChunkNotFoundException. 
Fixed it by making it optimistic check.